Validate forge packages

franciszekjob opened the pull request:
<!-- Reference any GitHub issues resolved by this PR -->

Closes #2975
Closes #2976

## Introduced changes

<!-- A brief description of the changes -->

- Validate `snforge_std` and `sncast_std` templates
- Validate (`scarb check` and `scarb lint`) on forge templates

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

ddoktorski committed:
Move init tests to `new.rs` file

ddoktorski committed:
Add Cairo program template

ddoktorski committed:
Update docs and changelog

ddoktorski committed:
Add `TemplateManifestConfig`

ddoktorski committed:
Add `erc20-contract` template

franciszekjob committed:
Fix packages linting

franciszekjob committed:
Add scripts for formatting and linting

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Disable running lint script

franciszekjob committed:
Comment out using asdf action

franciszekjob committed:
Fix forge debugging test

franciszekjob committed:
Restore previous CI changes

franciszekjob committed:
Remove print

franciszekjob committed:
Update gas calculation tests

franciszekjob committed:
Add prints

franciszekjob committed:
print diff

franciszekjob committed:
print diff

franciszekjob committed:
Update gas value in `contract_keccak_cost_sierra_gas`

franciszekjob committed:
Update gas value in `contract_keccak_cost_sierra_gas`

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Restore Cairo listings verification

franciszekjob commented on crates/forge/tests/integration/gas.rs:
note: Gas has changed because we're using `!=` instead of `<` comparison.

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into 2850-cairo-program-template

franciszekjob committed:
Merge branch '2850-cairo-program-template' of https://github.com/foundry-rs/starknet-foundry into 2850-erc-20-template

franciszekjob committed:
Merge branch '2850-erc-20-template' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Add `setup_snforge_packages.sh` script

franciszekjob committed:
Display output in llint script

franciszekjob committed:
Add comment in lint script

franciszekjob committed:
Fix permissions of `setup_snforge_packages.sh`

franciszekjob committed:
Add building forge

franciszekjob committed:
Refactor formatting and linting job

franciszekjob committed:
Format Cairo files

franciszekjob committed:
Update linting script

ddoktorski committed:
Move `add_dependencies` and `update_config` to `TemplateManifestConfig`

ddoktorski committed:
Add a note about template to Getting Started docs

ddoktorski committed:
Fix `assert_macros` version

ddoktorski committed:
Resolve merge conflicts

ddoktorski committed:
Use strk address instead of eth

ddoktorski committed:
Resolve merge conflicts

franciszekjob committed:
Remove asdf commands from `scarb_lint`

franciszekjob committed:
Fix CI workflow

franciszekjob committed:
Fix CI workflow

franciszekjob committed:
Use rust cache

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Fix linting in Cairo files

franciszekjob committed:
Add `allow-warnings` to `Scarb.toml` in default forge package

franciszekjob commented on .github/workflows/ci.yml:
```diff
@@ -267,22 +267,29 @@ jobs:
+        run: ./scripts/scarb_fmt.sh
+      - uses: software-mansion/setup-scarb@v1
+        with:
+          scarb-version: "2.10.0"
+      - name: Check packages linting
+        run: ./scripts/scarb_lint.sh
```
note: `scarb lint` is available from 2.10.0, hence need to use different scarb version

franciszekjob committed:
Merge branch '2850-erc-20-template' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Code cleanup

franciszekjob committed:
Update changelog

franciszekjob committed:
Update changelog

franciszekjob committed:
Remove displaying output in `setup_snforge_templates.sh`

franciszekjob committed:
Cleanup linting script

franciszekjob committed:
Fix tests

franciszekjob commented on crates/forge/tests/integration/gas.rs:
```diff
@@ -1201,15 +1201,15 @@ fn contract_keccak_cost_sierra_gas() {
-    // 1158935 = reported consumed sierra gas
-    // 0 l1_gas + 96 l1_data_gas + (142810 + 50000 + 87650 + 1158935) l2 gas
+    // 1157519 = reported consumed sierra gas
+    // 0 l1_gas + 96 l1_data_gas + (142810 + 50000 + 87650 + 1157519) l2 gas
     assert_gas(
         &result,
         "contract_keccak_cost",
         GasVector {
             l1_gas: GasAmount(0),
             l1_data_gas: GasAmount(96),
-            l2_gas: GasAmount(1_439_395),
+            l2_gas: GasAmount(1_438_075),
```
note: Gas changes are because of modifications in gas checker contract - now we're using `!=` instead of `<` in loops.

franciszekjob committed:
Fix CI

ddoktorski commented on scripts/scarb_lint.sh:
Tbh do we need checking lint in test packages at all? What are the benefits of this?

franciszekjob replied:
It's a tiny improvement of devex. It's not a big effort so imo we should leave it.

ddoktorski replied:
I must disagree. It seems like quite a lot of effort to maintain this compared to the value it brings, and it also makes our already fat CI even bigger.

franciszekjob committed:
Apply code review suggestions

franciszekjob commented on .github/workflows/ci.yml:
```diff
@@ -267,22 +267,29 @@ jobs:
-              exit 1
-          fi
-          exit 0
+          chmod +x ./scripts/setup_snforge_templates.sh
```
good catch 👍

franciszekjob commented on scripts/setup_snforge_templates.sh:
No, assuming that packages are correctly lined and can be compiled.

kkawula commented on scripts/scarb_lint.sh:
```diff
@@ -0,0 +1,43 @@
+
+echo "$output"
+
+# Check for lint warnings, we need to do it this way because `scarb lint` returns 0 even if there are warnings
+if grep -iq "warning: Plugin diagnostic:" <<< "$output"; then
+    exit 1
+# Check for missing `edition`` field
+elif grep -iq "warn: `edition` field not set in `[package]` section for package" <<< "$output"; then
+    exit 1
```
Maybe it is worth creating an issue for that. If in the future, `scarb lint` can change warning into an error

franciszekjob replied:
#3148

franciszekjob commented on CHANGELOG.md:
```diff
@@ -12,6 +12,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 #### Added
 
 - `--template` flag to `snforge new` command that allows selecting a template for the new project. Possible values are `balance-contract` (default), `cairo-program` and `erc20-contract`
+- `Scarb.toml` created by the `snforge new` command now includes the entry `allow-warnings = false` for Cairo checks
```
Fine, it may be too impactful. removing.

franciszekjob committed:
Refactor packages validation

franciszekjob committed:
Add packages validation tests

franciszekjob committed:
Refactor packages validation

franciszekjob committed:
Remove linting changes in internal packages

franciszekjob committed:
Little cleanup

franciszekjob commented:
@ddoktorski @kkawula
I've removed lint-related changes with internal packages and moved the tests from CI to e2e (as discussed with @ddoktorski ). Take a look

cptartur commented on .github/workflows/ci.yml:
```diff
@@ -267,23 +270,6 @@ jobs:
       - uses: software-mansion/setup-universal-sierra-compiler@v1
       - run: cargo test --release -p scarb-api
 
-  scarbfmt:
```
Any reason we no longer do this?

franciszekjob replied:
We do it in e2e test

cptartur commented on .github/workflows/ci.yml:
```diff
@@ -320,6 +306,8 @@ jobs:
       - uses: actions/checkout@v4
       - uses: software-mansion/setup-scarb@v1
+        with:
+          scarb-version: "2.10.0"
```
Why are we using a fixed scarb version here? 
Also I guess we should eventually remove this, an issue for this would be good.

franciszekjob replied:
Not needed, it was needed because of some linting changes (which aren't present anymore). Removed.

cptartur commented on crates/forge/tests/e2e/packages_validation.rs:
```diff
@@ -0,0 +1,149 @@
+use scarb_api::ScarbCommand;
+use test_case::test_case;
+
+fn find_files_recursively(dir_path: Utf8PathBuf, file_name: &str) -> Vec<Utf8PathBuf> {
+    let mut matches = Vec::new();
+    let paths = std::fs::read_dir(dir_path).unwrap();
+
+    for path in paths {
+        let path = path.unwrap().path();
+        if path.is_dir() {
+            matches.append(&mut find_files_recursively(
+                Utf8PathBuf::from(path.to_string_lossy().to_string()),
+                file_name,
+            ));
+        } else if path.to_string_lossy().contains(file_name) {
+            matches.push(Utf8PathBuf::from(path.to_string_lossy().to_string()));
+        }
+    }
```
I think we could use walkdir for this, no? https://docs.rs/walkdir/latest/walkdir/

cptartur commented on crates/forge/tests/e2e/packages_validation.rs:
```diff
@@ -0,0 +1,149 @@
+use camino::Utf8PathBuf;
+use clap::ValueEnum;
+use forge::Template;
+use project_root::get_project_root;
```
This package is just a single function with a couple of lines, maybe we could just implement it ourselves?

franciszekjob replied:
We can but don't wan to mix changes in this PR, can be addressed in separate one.

cptartur commented on crates/forge/tests/e2e/packages_validation.rs:
```diff
@@ -0,0 +1,149 @@
+        // Overwrite Scarb.toml with `allow-warnings = true`
+        let scarb_toml_path = package_path.join("Scarb.toml");
+        let mut scarb_toml = fs::read_to_string(&scarb_toml_path).unwrap();
+        scarb_toml.push_str("\n[cairo]\nallow-warnings = true\n");
```
Why do we allow them?

franciszekjob replied:
We do this only to ensure that scarb check will fail if there are any warnings (that's even suggested in issue description + we've talked about it offline yesterday 😅 ) . Note that we do it only for test packages, we don't include it in `Scarb.toml` template.

cptartur replied:
But shouldn't it be `allow_warnings=false` then?

franciszekjob replied:
Oops, should be, it was a typo.

cptartur commented on crates/forge/tests/e2e/packages_validation.rs:
```diff
@@ -0,0 +1,149 @@
+}
+
+#[test_case("snforge_std")]
+#[test_case("sncast_std")]
```
This validates sncast_std even though it is in `forge` crate. I think it'd make more sense in a separate crate that checks both of these crates (and we could expand it to other crates easily in the future if needed).

franciszekjob replied:
Good catch, will move it.

franciszekjob committed:
Remove scarb version from CI mdbbok step

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Set `allow-warnings = false` in template

franciszekjob committed:
Add `packages_validation` crate

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Fix newline

franciszekjob committed:
Code cleanup

franciszekjob committed:
Remove unneeded changes

franciszekjob committed:
Revert "Remove unneeded changes"

This reverts commit 9b3b1b165484ff078681e64fcde95253c5d6fe89.

franciszekjob committed:
Fix scarb formatting script

franciszekjob committed:
Update `scarbfmt.sh` script

franciszekjob committed:
Remove `exit 0` from `scarbfmt.sh` script

ddoktorski commented on .github/workflows/ci.yml:
```diff
@@ -140,6 +140,8 @@ jobs:
       - uses: software-mansion/setup-universal-sierra-compiler@v1
 
       - run: cargo test --package forge --features scarb_since_2_10 sierra_gas
+      - run: cargo test --package forge --features scarb_since_2_10 validate_templates
+      - run: cargo test --package packages_validation --features scarb_since_2_10 validate_cairo_libs
```
It would be better to extend `scarbfmt` job and add these steps there. It could be renamed to something like `scarb fmt and lint`

franciszekjob replied:
What's the rationale? Packages validation tests are run twice - normally in e2e suite and here with scarb 2.10.
I don't think we should spread tests run with `scarb_since_2_10` feature flag across many jobs.
Once our minimal supported scarb version is 2.10, then we will just remove this job and there won't be need to modify others.

franciszekjob replied:
> That's true in the case of validate_templates, but what about the packages_validation tests?

Right, it should be added to run without feature flag, now I see your point.

ddoktorski commented on crates/forge/tests/e2e/templates_validation.rs:
```diff
@@ -0,0 +1,41 @@
+
+    let templates = Template::value_variants();
+
+    for template in templates {
```
nit: Why not to use `#[test_case]` instead for each template?

franciszekjob replied:
To prevent from forgetting about adding test case when new template is added.

kkawula commented on scripts/scarbfmt.sh:
```diff
@@ -1,12 +1,10 @@
-' \;
-
-popd || exit
+output=$(find . -type f -name "Scarb.toml" -execdir sh -c '
+    echo "Running \"scarb fmt --check\" in directory: $PWD"
+    scarb fmt --check
+' \;)
+echo "$output"
+if grep -iq "Diff" <<< "$output"; then
+    exit 1
+fi
```
nit. bash generally uses 2 spaces for indentation
you can easily format code with `shfmt -i 2 <file>`
```suggestion
output=$(find . -type f -name "Scarb.toml" -execdir sh -c '
    echo "Running \"scarb fmt --check\" in directory: $PWD"
    scarb fmt --check
' \;)
echo "$output"
if grep -iq "Diff" <<<"$output"; then
  exit 1
fi
```

franciszekjob committed:
Apply code review suggestion to CI

franciszekjob committed:
Fix CI

franciszekjob committed:
Formatting

franciszekjob committed:
Merge branch 'master' into franciszekjob/2975-lint-packages

franciszekjob committed:
Fix CI

franciszekjob committed:
Merge branch 'franciszekjob/2975-lint-packages' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Run `shfmt -i 2` in `scarbfmt.sh`

franciszekjob committed:
Fix CI

ddoktorski commented on crates/packages_validation/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "packages_validation"
+version = "1.0.0"
+edition.workspace = true
```
Sounds good to me

franciszekjob replied:
as dicsussed offline: xtask seems good, but in order to validate templates we're using enum from forge, also the `check_and_lint` function is used both in libs and templates validation. Let's leave it in crate.

cptartur commented:
@franciszekjob please re-request when it's ready for another review (xtask is implemented?)

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Merge branch 'master' into franciszekjob/2975-lint-packages

kkawula commented on crates/packages_validation/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "packages_validation"
```
Maybe it should be renamed? 
sth like `testing_packages_validatoin`, in order to distinguish this package from others

franciszekjob replied:
I think current name already distinguishes it from others, but I'm open for change. Wdyt @ddoktorski @cptartur ?

cptartur replied:
Does rust workspace structure allow for nesting? If so, maybe it'd make sense to have a `crates/testing` directory? Then we wouldn't need to include `testing_` in the crate name.

We could possible move code from shared::test_utils there as well.

franciszekjob replied:
I've added new crates `testing` and `packages_validation` inside. 

> We could possible move code from shared::test_utils there as well.

Will see it that makes sense and potentially move it in separate PR 👍

franciszekjob commented on crates/packages_validation/src/lib.rs:
```diff
@@ -0,0 +1,37 @@
+use std::process::Stdio;
+
+pub fn check_and_lint(package_path: Utf8PathBuf) {
+    println!("Running `scarb check` in directory {package_path}");
```
Without this we won't know for which package `scarb check` has failed - it's not cluttering.
We do the same for docs validation btw.

franciszekjob commented on scripts/scarbfmt.sh:
In fact, it's in separate script 😅 . Am I missing something?

franciszekjob replied:
Restored.

franciszekjob commented on crates/packages_validation/src/lib.rs:
```diff
@@ -0,0 +1,37 @@
+    assert!(check_output.status.success(), "`scarb check` failed");
+
+    // TODO(#3149)
+    if cfg!(feature = "scarb_since_2_10") {
```
Could you clarify a bit? We run `scarb lint` only for scarb 2.10.0 (because it provides lint).

franciszekjob replied:
discussed offline, it's fine as it is now

cptartur commented:
@franciszekjob please re-request review after outstanding comments have been answered (or addressed) 🙏

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Merge branch 'franciszekjob/2975-lint-packages' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Fix formatting

franciszekjob committed:
Add `scarb_since_2_10` feature flag in `sncast`

franciszekjob committed:
Update forge `Cargo.toml`

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Move running check and lint tests to different jobs

franciszekjob committed:
Rename scarb fmt job

franciszekjob commented on crates/packages_validation/tests/cairo_libs.rs:
Moved

franciszekjob committed:
Rename scarb fmt job

franciszekjob committed:
Perform self code review

franciszekjob committed:
Rename scarb fmt job

franciszekjob committed:
Restore CI steps

franciszekjob committed:
Restore scarbfmt script

franciszekjob committed:
Remove formatting changes

franciszekjob committed:
Merge branch 'master' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2975-lint-packages

franciszekjob committed:
Update `Cargo.lock`

franciszekjob commented on .github/workflows/ci.yml:
```diff
@@ -276,6 +284,12 @@ jobs:
       - name: Run tests
         run: cargo test --release -p sncast
+      # TODO(#3096): Remove running tests with `scarb_since_2_10`
+      - uses: software-mansion/setup-scarb@v1
+        with:
+          scarb-version: "2.10.1"
```
But we need to set specific version of scarb here, in this case 2.10.1, that's why we use it again.

kkawula replied:
We probably can removed one of these

franciszekjob replied:
No, because then we will use version from `.tool-versions` file, which ATM is 2.9.4.
Once minimal supported is 2.10, then we can remove this step (as mentioned in the comment).

franciszekjob commented on .github/workflows/ci.yml:
```diff
@@ -44,6 +44,14 @@ jobs:
       - uses: software-mansion/setup-universal-sierra-compiler@v1
       - run: cargo test --release --lib -p forge
+      # TODO(#3096): Remove running tests with `scarb_since_2_10`
+      - uses: software-mansion/setup-scarb@v1
+        with:
+          scarb-version: "2.10.1"
```
Reply same as above.

franciszekjob committed:
Refactor `validate_templates` test

kkawula commented on crates/forge/tests/e2e/templates.rs:
```diff
@@ -0,0 +1,38 @@
+
+    // Overwrite Scarb.toml with `allow-warnings = false`
+    let scarb_toml_path = package_path.join("Scarb.toml");
+    let mut scarb_toml = fs::read_to_string(&scarb_toml_path).unwrap();
+    scarb_toml.push_str("\n[cairo]\nallow-warnings = false\n");
+    fs::write(&scarb_toml_path, &scarb_toml).expect("Failed to write to Scarb.toml");
```
Maybe to avoid bugs in the future, we should use `toml_edit` crate to add this entry without creating conflicts?

cptartur replied:
+1 for `toml_edit`, we already use it in some places.

franciszekjob replied:
Applied

cptartur commented on Cargo.toml:
```diff
@@ -14,7 +14,9 @@ members = [
     "crates/docs",
-    "crates/debugging"
+    "crates/debugging",
+    "crates/testing",
+    "crates/testing/packages_validation",
```
Can't we do something along the lines of this?

```suggestion
    "crates/testing/*"
```

franciszekjob replied:
I don't think so. Note that above we do the same:

https://github.com/foundry-rs/starknet-foundry/blob/master/Cargo.toml#L9-L10

cptartur commented on crates/testing/Cargo.toml:
I don't think we need an empty `testing` package (unless there is now way around it). It should just be a plain directory containing testing related crates.

cptartur commented on crates/testing/packages_validation/src/lib.rs:
```diff
@@ -0,0 +1,54 @@
+    }
+}
+
+pub fn validate_cairo_lib(package_name: &str) {
+    let package_path = project_root::get_project_root()
+        .expect("Failed to get project root")
+        .join(package_name);
+    let package_path = package_path
+        .canonicalize()
+        .expect("Failed to canonicalize path");
+    let package_path =
+        Utf8PathBuf::from_path_buf(package_path).expect("Failed to convert to Utf8PathBuf");
+
+    check_and_lint(&package_path);
+}
```
Not sure if resolving paths to packages should be responsibility of this package - it makes it less flexible. Why not provide paths at the caller site? 

The paths when running `cargo test` have a stable working directory, so we can just hardcode them at caller site https://doc.rust-lang.org/cargo/commands/cargo-test.html#working-directory-of-tests

franciszekjob replied:
> Not sure if resolving paths to packages should be responsibility of this package - it makes it less flexible. Why not provide paths at the caller site?

Valid point, applied

cptartur commented on crates/forge/tests/e2e/templates.rs:
```diff
@@ -0,0 +1,38 @@
+use packages_validation::check_and_lint;
+use test_case::test_case;
+
+#[test_case(&Template::CairoProgram; "cairo-program")]
+#[test_case(&Template::BalanceContract; "balance-contract")]
+#[test_case(&Template::Erc20Contract; "erc20-contract")]
+fn validate_templates(template: &Template) {
+    let temp_dir = assert_fs::TempDir::new().expect("Unable to create a temporary directory");
+    let package_name = format!("{}_test", template.to_string().replace('-', "_"));
+
+    runner(&temp_dir)
+        .args([
+            "new",
+            "--template",
+            template.to_string().as_str(),
+            &package_name,
+        ])
+        .assert()
+        .success();
+
+    let package_path = temp_dir.child(package_name);
+    let package_path = Utf8PathBuf::from_path_buf(package_path.to_path_buf())
+        .expect("Failed to convert to Utf8PathBuf");
```
We can probably hardcode these, see my comment about paths below

franciszekjob replied:
In this case w we shouldn't hardcode them - note that these the package name is dynamic, depending on specific test case.

cptartur replied:
Test case can have a reference to the template and a path to it. But maybe that's more complicated than the current solution

franciszekjob replied:
I would stay with current one tbh.

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Update members in `Cargo.toml`

franciszekjob committed:
Apply code review suggestions

kkawula commented on crates/forge/tests/code_quality.rs:
```diff
@@ -0,0 +1,18 @@
+use camino::Utf8PathBuf;
+use packages_validation::check_and_lint; // Ensure you have this at the top
+
```
What is the idea behind this?

franciszekjob replied:
Leftover, removed 😅

franciszekjob committed:
Remove comment

cptartur commented on crates/forge/tests/code_quality.rs:
```diff
@@ -0,0 +1,18 @@
+        .canonicalize()
+        .expect("Failed to canonicalize path")
+        .try_into()
+        .expect("Failed to convert to Utf8PathBuf");
```
Why not just provide this path directly if the working directory of tests is stable?

```suggestion
    let package_path = Utf8PathBuf::from("../../snforge_std")
        .canonicalize()
        .unwrap()
        .try_into()
        .unwrap();
```

cptartur commented on crates/sncast/tests/code_quality.rs:
```diff
@@ -0,0 +1,18 @@
+
+#[test]
+fn validate_sncast_std() {
+    let package_path: Utf8PathBuf = std::env::current_dir()
```
Same here

franciszekjob committed:
Apply code review suggestion
